sql/json_value.go

```diff
@@ -0,0 +1,371 @@
+// Copyright 2021 Dolthub, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package sql
+
+import (
+	"encoding/json"
+	"github.com/oliveagle/jsonpath"
+	"sort"
+	"strings"
+)
+
+// JSONValue is an integrator specific implementation of a JSON field value.
+type JSONValue interface {
+	// Unmarshall converts a JSONValue to a JSONDocument
+	Unmarshall() (val JSONDocument, err error)
+	// Compare compares two JSONValues. It maintains the same return value
+	// semantics as Type.Compare()
+	Compare(v JSONValue) (cmp int, err error)
+	// ToString marshalls a JSONValue to a valid JSON-encoded string.
+	ToString() (string, error)
+}
+
+// SearchableJSONValue is JSONValue supporting in-place access operations.
+// The query engine can utilize these optimized access methods improve performance
+// by minimizing the need to unmarshall a JSONValue into a JSONDocument.
+type SearchableJSONValue interface {
```

Comment:
Should have method doc for these interface methods. Fine to sketch out if they're not finalized